File: Windows/universal8.1/Siminov/Core/Events/IDatabaseEvents.cs
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/



using Siminov.Core.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Siminov.Core.Events
{

    /// <summary>
    /// Exposes methods which deal with events associated with database operation's.
    /// It has methods such as (databaseCreated, databaseDroped, tableCreated, tableDroped, indexCreated).
    /// </summary>
    public interface IDatabaseEvents
    {

        /// <summary>
        /// This event is fired when database gets created as per database descriptor.
        /// </summary>
        /// <param name="databaseDescriptor">contains meta data associated with database</param>
        void OnDatabaseCreated(DatabaseDescriptor databaseDescriptor);

        /// <summary>
        /// This event is fired when database is dropped.
        /// </summary>
        /// <param name="databaseDescriptor">contains meta data associated with dropped database</param>
        void OnDatabaseDropped(DatabaseDescriptor databaseDescriptor);



        /// <summary>
        /// This event is fired when a table is created.
        /// </summary>
        /// <param name="databaseDescriptor">Database descriptor object</param>
        /// <param name="entityDescriptor">Entity descriptor object</param>
        void OnTableCreated(DatabaseDescriptor databaseDescriptor, EntityDescriptor entityDescriptor);

        /// <summary>
        /// This event is fired when a table is dropped.
        /// </summary>
        /// <param name="databaseDescriptor">Database descriptor object</param>
        /// <param name="entityDescriptor">Entity descriptor object</param>
        void OnTableDropped(DatabaseDescriptor databaseDescriptor, EntityDescriptor entityDescriptor);



        /// <summary>
        /// This event is fired when a index is created on table.
        /// </summary>
        /// <param name="databaseDescriptor">Database descriptor object</param>
        /// <param name="entityDescriptor">Entity descriptor object</param>
        /// <param name="index">meta data about index got created</param>
        void OnIndexCreated(DatabaseDescriptor databaseDescriptor, EntityDescriptor entityDescriptor, EntityDescriptor.Index index);



        /// <summary>
        /// This event is fired when a index is dropped.
        /// </summary>
        /// <param name="databaseDescriptor">Database descriptor object</param>
        /// <param name="entityDescriptor">Entity descriptor object</param>
        /// <param name="index">meta data about index got dropped</param>
        void OnIndexDropped(DatabaseDescriptor databaseDescriptor, EntityDescriptor entityDescriptor, EntityDescriptor.Index index);

    }
}

File: Android/Core/src/main/java/siminov/core/events/IDatabaseEvents.java
/**
 * [SIMINOV FRAMEWORK - CORE]
 * Copyright [2014-2016] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/


package siminov.core.events;

import siminov.core.model.DatabaseDescriptor;
import siminov.core.model.EntityDescriptor;
import siminov.core.model.EntityDescriptor.Index;

/**
 * Exposes methods which deal with events associated with database operation's.
 * It has methods such as (databaseCreated, databaseDroped, tableCreated, tableDroped, indexCreated).
 */
public interface IDatabaseEvents {

	/**
	 * This event is fired when database gets created as per database descriptor.
	 * @param databaseDescriptor contains meta data associated with database.
	 */
	public void onDatabaseCreated(final DatabaseDescriptor databaseDescriptor);
	
	/**
	 * This event is fired when database is dropped.
	 * @param databaseDescriptor contains meta data associated with dropped database.
	 */
	public void onDatabaseDropped(final DatabaseDescriptor databaseDescriptor);
	
	
	
	/**
	 * This event is fired when a table is created.
	 * @param entityDescriptor contains meta data associated with created table.
	 */
	public void onTableCreated(final DatabaseDescriptor databaseDescriptor, final EntityDescriptor entityDescriptor);
	
	/**
	 * This event is fired when a table is dropped.
	 * @param entityDescriptor contains meta data associated with dropped table.
	 */
	public void onTableDropped(final DatabaseDescriptor databaseDescriptor, final EntityDescriptor entityDescriptor);
	
	
	
	
	/**
	 * This event is fired when a index is created on table.
	 * @param entityDescriptor contains meta data associated with table on which index is created.
	 * @param index meta data about index got created.
	 */
	public void onIndexCreated(final DatabaseDescriptor databaseDescriptor, final EntityDescriptor entityDescriptor, Index index);
	
	/**
	 * This event is fired when a index is dropped.
	 * @param entityDescriptor contains meta data associated with table on which index is dropped.
	 * @param index meta data about index got dropped.
	 */
	public void onIndexDropped(final DatabaseDescriptor databaseDescriptor, final EntityDescriptor entityDescriptor, Index index);
	
}
